CMakeLists.txt

```diff
@@ -150,7 +150,7 @@ if(${MM_MALLOC})
 endif()
 
 if(UNIX OR MINGW OR CYGWIN)
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -pthread -O3 -Wextra -Wall -Wno-ignored-attributes -Wno-unknown-pragmas -Wno-return-type")
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -pthread -O3 -Wextra -Wall -Wno-ignored-attributes -Wno-unknown-pragmas -Wno-return-type")
```

Comment:
To support wide range of OSes we target gcc 4.8.2. Unfortunately, C++-17 is supported only starting from gcc 5.
Refer to 
https://github.com/microsoft/LightGBM/blob/704e08300ead3e70625d5431c845141f2e564123/CMakeLists.txt#L20-L38

Reply:
So I assume this change is currently impossible, any plan to migrate?

Reply:
I think there are two options:
1) implement shared_mutex by our own. But this seems untrivial. 
2) define a preprocessor, called `DUSE_CPP17`, like `DUSE_MPI`, in cmake, and add `-std=c++17` if DUSE_CPP17 is set in cmake. And when `DUSE_CPP17` is used, use shared_mutex in cpp code.

Reply:
Hello @guolinke thanks for the proposal , I did some changes with this proposed implementation